dpnp/backend/kernels/dpnp_krnl_reduction.cpp

```diff
@@ -114,47 +114,90 @@ void dpnp_sum_c(void* result_out,
         // type of "init" determine internal algorithm accumulator type
         _DataType_output accumulator = std::reduce(
             policy, input_it.begin(output_id), input_it.end(output_id), init, std::plus<_DataType_output>());
-        policy.queue().wait();
+        policy.queue().wait(); // TODO move out of the loop
 
         result[output_id] = accumulator;
     }
 
     return;
 }
 
-template <typename _KernelNameSpecialization>
+template <typename _KernelNameSpecialization1, typename _KernelNameSpecialization2>
 class dpnp_prod_c_kernel;
 
-template <typename _DataType>
-void dpnp_prod_c(void* array1_in, void* result1, size_t size)
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_prod_c(void* result_out,
+                 const void* input_in,
+                 const size_t* input_shape,
+                 const size_t input_shape_ndim,
+                 const long* axes,
+                 const size_t axes_ndim,
+                 const void* initial, // type must be _DataType_output
+                 const long* where)
 {
-    if (!size)
+    (void)where; // avoid warning unused variable
+
+    if ((input_in == nullptr) || (result_out == nullptr))
     {
         return;
     }
 
-    _DataType* array_1 = reinterpret_cast<_DataType*>(array1_in);
-    _DataType* result = reinterpret_cast<_DataType*>(result1);
+    const _DataType_output init = get_initial_value<_DataType_output>(initial, 1);
+
+    _DataType_input* input = get_array_ptr<_DataType_input>(input_in);
+    _DataType_output* result = get_array_ptr<_DataType_output>(result_out);
```

Comment:
We can avoid `const_cast` for `result_out`. Wouldn't we like to move `get_array_ptr` to utils maybe in separate PR?

Reply:
"yes" and "yes" :-)
`const_cast` for `result` has no meaning but I use the function `get_array_ptr` to consistency.